src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -745,6 +746,23 @@ private bool CheckFieldValueKind(SyntaxNode node, BoundFieldAccess fieldAccess,
 
             // for other fields defer to the receiver.
             return CheckIsValidReceiverForVariable(node, fieldAccess.ReceiverOpt, valueKind, diagnostics);
+
+            bool isAssignedFromInitOnlySetterOnThis(BoundExpression receiver)
+            {
+                // bad: other.readonlyField = ...
+                // bad: base.readonlyField = ...
+                if (!(receiver is BoundThisReference))
```

Comment:
>if (!(receiver is BoundThisReference)) [](start = 16, length = 38)

It looks like the speclet doesn't have this restriction. Is only says: "Assign readonly fields declared on the same type". However, I think the intention was to have this restriction. Please make sure to reconcile the difference.

Reply:
Updated speclet

---
In reply to: [415209113](https://github.com/dotnet/roslyn/pull/43275#discussion_r415209113) [](ancestors = 415209113)